packages/cli/src/utils/tunnel-utils.ts

```diff
@@ -0,0 +1,141 @@
+    if (event.type === 'error') {
+      return 'error'
+    }
+
+    if (event.type === 'close') {
+      return `${event.ev.code} ${event.ev.reason}`
+    }
+
+    return 'init'
+  }
+}
+
+export class TunnelSupervisor {
+  private _tunnel?: TunnelTail
+  private _closed = false
+  public readonly events = new EventEmitter<{
+    connectionFailed: ReconnectionTriggerEvent
+    manuallyClosed: null
+    connected: {
+      tunnel: TunnelTail
+      ev: ReconnectionTriggerEvent
+    }
+  }>()
+
+  constructor(private _tunnelUrl: string, private _tunnelId: string, private _logger: Logger) {}
+
+  public async start(): Promise<void> {
+    if (this._closed) {
+      throw new Error('Tunnel is closed')
+    }
+    if (this._tunnel) {
+      throw new Error('Tunnel is already started')
+    }
+
+    const tunnel = await this._reconnect({ type: 'init', ev: null })
+    this._tunnel = tunnel
+  }
+
+  public get closed(): boolean {
+    return this._closed
+  }
+
+  /**
+   * @returns Promise that rejects when a reconnection attempt fails and resolves when the tunnel is closed manually
+   */
+  public async wait(): Promise<void> {
+    if (this._closed) {
+      throw new Error('Tunnel is closed')
+    }
+
+    return new Promise((resolve, reject) => {
+      this.events.on('connectionFailed', (ev) => {
+        reject(new ReconnectionFailedError(ev))
+      })
+
+      this.events.on('manuallyClosed', () => {
+        resolve()
+      })
+    })
+  }
+
+  public close(): void {
+    if (this._closed) {
+      return
+    }
+
+    this._closed = true
+    this._tunnel?.close()
+    this.events.emit('manuallyClosed', null)
+  }
+
+  private _reconnectSync(ev: ReconnectionTriggerEvent): void {
+    void this._reconnect(ev)
+      .then((t) => {
+        this._tunnel = t
+      })
+      .catch(() => this.events.emit('connectionFailed', ev))
+  }
+
+  private async _reconnect(ev: ReconnectionTriggerEvent): Promise<TunnelTail> {
+    const newTunnel = async () => {
+      const tunnel = await TunnelTail.new(this._tunnelUrl, this._tunnelId)
+      this._registerListeners(tunnel)
+      this.events.emit('connected', { tunnel, ev })
+      return tunnel
+    }
+
+    if (ev.type === 'init') {
+      // skip logging on the first connection attempt
+      return newTunnel()
+    }
+
+    const line = this._logger.line()
+    line.started('Reconnecting tunnel...')
+    const tunnel = await newTunnel()
+    line.success('Reconnected')
+    line.commit()
+    return tunnel
+  }
+
+  private _registerListeners(tunnel: TunnelTail) {
+    tunnel.events.on('error', ({ target, type }) => {
+      this._logger.error(`Tunnel error: ${type}`)
+      this._reconnectSync({ type: 'error', ev: { target, type } })
+    })
+    tunnel.events.on('close', ({ code, reason, target, type, wasClean }) => {
+      this._logger.error(`Tunnel closed: ${code} ${reason}`)
+      this._reconnectSync({ type: 'close', ev: { code, reason, target, type, wasClean } })
```

Comment:
**❔ question:** ‏ What is the `wasClean` flag used for?

Reply:
I'm not sure, it comes the native web socket close event. I think its about weither or not the connection was closed cleanly. In this case, I only deconstruct it to pass it around below